### PR TITLE
DATAJPA-1333 - Minor performance improvements

### PR DESCRIPTION
**Pull request by stsypanov**
PR for [https://jira.spring.io/browse/DATAJPA-1333](https://jira.spring.io/browse/DATAJPA-1333)

### CODE REVIEW

File: src/main/java/org/springframework/data/jpa/convert/QueryByExamplePredicateBuilder.java [hunk 1]
```diff
@@ -94,7 +97,7 @@
 			return predicates.iterator().next();
 		}
 
-		Predicate[] array = predicates.toArray(new Predicate[predicates.size()]);
+		Predicate[] array = predicates.toArray(new Predicate[0]);
```
[the remark this answers is not in the document]
Comment (lukaseder) [reply src/main/java/org/springframework/data/jpa/convert/QueryByExamplePredicateBuilder.java]: Nope: https://shipilev.net/blog/2016/arrays-wisdom-ancients/

[the remark this answers is not in the document]
Comment (stsypanov) [reply src/main/java/org/springframework/data/jpa/convert/QueryByExamplePredicateBuilder.java]: Actually as mentioned in https://shipilev.net/blog/2016/arrays-wisdom-ancients/ passing of zero-length array allows VM dodge zeroing elimination and do some more magic.

Passing `size` into array constructor is an outdated recommendation. It's reverted in modern IDEs, see e.g. https://youtrack.jetbrains.com/issue/IDEA-180847

[the remark this answers is not in the document]
Comment (schauder) [reply src/main/java/org/springframework/data/jpa/convert/QueryByExamplePredicateBuilder.java]: Interesting. Of course, that leaves me even more hesitant to change the code for performance reasons, when even such "obvious cases" aren't obvious at all.

[the remark this answers is not in the document]
Comment (stsypanov) [reply src/main/java/org/springframework/data/jpa/convert/QueryByExamplePredicateBuilder.java]: Sometimes performance related advices become outdated. E. g. previously (I'm not sure was it before Java 7 or 8) when we wrote smth like
```java
String str = "abc.def";
String[] split = str.split(".");
```

IDEA would highligh `split` and propose to replace it with precompiled regexp. Later `java.lang.String::split` was changed and a fast path added for single-character strings where single char is one of the RegEx's meta characters ".$|()[{^?*+\\". After the change was introduced proposed replacement would bring pessimization so it's not recommended anymore.

I've benchmarked all changes proposed in this PR. All of them bring optimization either in footprint (e.g. `Arrays::asList` -> `Collections::singletonList`) or execution time (`Stream::anyMatch` -> `Collection::contains`) or even both of them (`HashSet -> EnumSet`)

[the remark this answers is not in the document]
Comment (schauder) [reply src/main/java/org/springframework/data/jpa/convert/QueryByExamplePredicateBuilder.java]: Cool. Can you share the benchmarks? Sounds like an opportunity to learn for me.

[the remark this answers is not in the document]
Comment (stsypanov) [reply src/main/java/org/springframework/data/jpa/convert/QueryByExamplePredicateBuilder.java]: Benchmark for `Collection::toArray` can be found in Alexey's article via the link above.

As for my changes proposed in this pull-request:

- case for `Stream::anyMatch` -> `Collection::contains`, see my [AnyMatchVsContainsBenchmark](https://github.com/stsypanov/logeek-night-benchmark/blob/master/benchmark-runners/src/main/java/com/luxoft/logeek/benchmark/collection/AnyMatchVsContainsBenchmark.java)

- case for unnecessary boxing (proposed changes into SimpleJpaRepository), see my [BoxingBenchmark](https://github.com/stsypanov/logeek-night-benchmark/blob/master/benchmark-runners/src/main/java/com/luxoft/logeek/benchmark/boxing/BoxingBenchmark.java)

To run them you can checkout https://github.com/stsypanov/logeek-night-benchmark, then find `BenchmarkRunner` and run it's `main` method. If memory and/or GC benchmarking required then uncomment line with `addProfile`

[the remark this answers is not in the document]
Comment (stsypanov) [reply src/main/java/org/springframework/data/jpa/convert/QueryByExamplePredicateBuilder.java]: @schauder ping

**Comment by schauder** [comment]
Thanks, that's merged.

**Pull request closed without merging**